### PR TITLE
[162] Build "Add a trainee" overview page

### PR DESCRIPTION
**Pull request by aliuk2012**
### Changes proposed in this pull request
- Adds new TaskList component
- Refactors the trainees show page to use the TaskList component and bring it inline with prototype

### Guidance to review

- Go to https://dfe-twd-rttd-pr-58.herokuapp.com/trainees , Click on the "Add data..." button, complete the page and see the overview page after clicking on "Continue"
- View existing record https://dfe-twd-rttd-pr-58.herokuapp.com/trainees/1
- View the new TaskList Component - 2 examples can be found by clicking on these links https://dfe-twd-rttd-pr-58.herokuapp.com/view_components/task_list/view

### CODE REVIEW

File: app/components/task_list/style.scss [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+.app-task-list__key {
+  @include govuk-responsive-margin(1, "right");
+}
+
+@media print {
+  .app-task-list__task-name::after {
+    display: none;
+  }
```
Comment (aliuk2012) [inline app/components/task_list/style.scss:46]: @edwardhorsford Should we allow the status to be included in print outs?

**Comment by aliuk2012** [comment]
[EPIC: Build Add a Trainee Overview](https://trello.com/c/hzKGoOLz/162-epic-build-add-a-trainee-overview)

File: app/components/task_list/style.scss (hunk 1 above)
Comment (fofr) [reply app/components/task_list/style.scss:46]: I'd expect a status to show when printed. Seems pretty important to the task list pattern.

Comment (edwardhorsford) [reply app/components/task_list/style.scss:46]: I see no reason not to.

File: app/views/trainees/show.html.erb [hunk 2]
```diff
@@ -1,59 +1,86 @@
...
+  <span class="govuk-caption-xl">Add a trainee</span>
+  Overview
+</h1>
 
-<div class="personal-details">
-  <%= render_component(SummaryCard::View.new(title: "Personal details",
-                                             rows: TraineePersonalDetails.new(@trainee).call))  %>
-</div>
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
```
Comment (edwardhorsford) [inline app/views/trainees/show.html.erb:15]:
```suggestion
  <div class="govuk-grid-column-two-thirds-from-desktop">
```

File: app/views/trainees/show.html.erb [hunk 3]
```diff
@@ -1,59 +1,86 @@
...
+    text: 'All records',
+    href: trainees_path
+  ) %>
+<% end %>
 
-<p class="govuk-body">
-  <%= govuk_link_to "Update", personal_details_trainee_path(@trainee) %>
-</p>
+<h1 class="govuk-heading-l">
+  <span class="govuk-caption-xl">Add a trainee</span>
```
Comment (edwardhorsford) [inline app/views/trainees/show.html.erb:10]: Should be `govuk-caption-l`

I'm not convinced this needs to be within the h1.

**Review by edwardhorsford — CHANGES_REQUESTED** [review]
Looking good, made a few suggestions.

It also needs a page title - suggest copying the helpers used on apply.

File: app/views/trainees/show.html.erb [hunk 4]
```diff
@@ -1,59 +1,86 @@
...
 </div>
 
-<h2 class="govuk-heading-l">Training details</h2>
-
-<p class="govuk-body">
-  <%= govuk_link_to "Update", training_details_trainee_path(@trainee) %>
-</p>
+<%= form_with url: trn_submissions_path, method: :post, local: true do |f| %>
+  <%= hidden_field_tag :trainee_id, @trainee.id %>
+  <%= f.govuk_submit "Check and submit for TRN" %>
```
Comment (edwardhorsford) [inline app/views/trainees/show.html.erb:83]: This renders out as an input - should it be a button?

File: app/views/trainees/show.html.erb (hunk 3 above)
Comment (aliuk2012) [reply app/views/trainees/show.html.erb:10]: Good catch but I think the caption needs to be inside the h1 to give screen reader users more context about what "Overview" is if they are using headings for in page navigation

Comment (edwardhorsford) [reply app/views/trainees/show.html.erb:10]: Ok, sounds good!

File: spec/components/task_list/view_preview.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+class TaskList::ViewPreview < ViewComponent::Preview
+  def default_state
+    render_component TaskList::View.new do |component|
+      component.slot(
+        :row,
+        task_name: "Personal",
```
Comment (edwardhorsford) [inline spec/components/task_list/view_preview.rb:6]:
```suggestion
        task_name: "Personal details",
```

File: app/views/trainees/show.html.erb [hunk 6]
```diff
@@ -1,59 +1,86 @@
...
-</div>
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <%= render_component TaskList::View.new(classes: "record-setup") do |component|
+      component.slot(
+        :row,
+        task_name: 'Record setup',
+        path: "",
+        status: 'completed'
+      )
```
Comment (misaka) [inline app/views/trainees/show.html.erb:22]: I'm still wrapping my head around view components and the govuk bits, but isn't the whole point of using `wrap_slot :row` to be able to just this:

```suggestion
      row task_name: 'Record setup',
          path: "",
          status: 'completed'
```

File: app/components/task_list/view.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
+class TaskList::View < GovukComponent::Base
+  include ViewComponent::Slotable
+
+  with_slot :row, collection: true, class_name: "Row"
+  wrap_slot :row
```
Comment (misaka) [inline app/components/task_list/view.rb:5]: Interesting. Did some looking into this, and this whole `wrap_slot` is giving me the chills. It [redefines `#row` with a method that just calls `component.slot(:row, ...)`](https://github.com/DFE-Digital/govuk-components/blob/0a9fb7391674b977081be1c912268716415edac4/app/components/govuk_component/base.rb#L10-L16). This overrides [the accessor that was defined by ViewComponent::Slotable with](https://github.com/github/view_component/blob/828714eaaeb62fbca8928ce588725d3ca2e7a2ff/lib/view_component/slotable.rb#L44-L53), as above, something that calls `component.slot(:row, ...)`. That method, unlike the accessor that was just wiped out, [returns nil](https://github.com/github/view_component/blob/828714eaaeb62fbca8928ce588725d3ca2e7a2ff/lib/view_component/slotable.rb#L118).

TL;DR: `wrap_slot` as-is breaks the `row` accessor (this issues is discussed here: https://github.com/github/view_component/discussions/454#discussioncomment-79355)

I'll make a comment on `govuk-components`.

Comment (misaka) [reply app/components/task_list/view.rb:5]: https://github.com/DFE-Digital/govuk-components/issues/63

Comment (misaka) [reply app/components/task_list/view.rb:5]: Seeing as the "wrapped" slot isn't used below, I'd suggest removing the `wrap_slot` for now.

**Comment by misaka** [comment]
[[M] Update Trainee Overview page](https://trello.com/c/iPHpnu0i/163-m-update-trainee-overview-page)

File: app/components/task_list/view.rb (hunk 7 above)
Comment (aliuk2012) [reply app/components/task_list/view.rb]: I've removed that line now

File: app/views/trainees/show.html.erb (hunk 4 above)
Comment (aliuk2012) [reply app/views/trainees/show.html.erb:83]: good spot, more of a issue for `govuk_design_system_formbuilder`

Comment (aliuk2012) [reply app/views/trainees/show.html.erb:83]: https://github.com/DFE-Digital/govuk_design_system_formbuilder/issues/203

**Review by defong — APPROVED** [review]

File: spec/features/trainees/creating_a_new_trainee_spec.rb [hunk 8]
```diff
@@ -38,6 +38,6 @@ def and_i_save_the_form
 
   def then_i_should_see_the_new_trainee_details
     expect(@show_page).to be_displayed
-    expect(@show_page).to have_content("Trainee ID 123")
+    # expect(@show_page).to have_content("Trainee ID 123")
```
Comment (misaka) [inline spec/features/trainees/creating_a_new_trainee_spec.rb:41]: Is it worth removing this?

File: spec/components/task_list/view_spec.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      it "renders a list of tasks" do
+        expect(component).to have_selector(".app-task-list__item")
+      end
+
+      it "renders the task name" do
+        expect(component).to have_link("some key", href: "some_path")
+      end
+    end
+
+    context "tags" do
```
Comment (misaka) [inline spec/components/task_list/view_spec.rb:30]: Minor quibble: we are testing the `tags` here (I believe?), so probably should be `describe "tags"`.

File: spec/components/task_list/view_spec.rb [hunk 10]
```diff
@@ -0,0 +1,68 @@
+require "rails_helper"
+
+RSpec.describe TaskList::View do
+  alias_method :component, :page
```
Comment (aliuk2012) [inline spec/components/task_list/view_spec.rb:4]: Love this!!! `page` in this context is a Capybara object of just the rendered component. so to improve the test @darokel helped me alias this so that its obvious that its just the component being tested. We will need to see if theres some way to make this a global option for component tests

Comment (misaka) [reply spec/components/task_list/view_spec.rb:4]: Ya, when I saw the `have_selector` stuff below it made me wonder if we can't use `SitePrism` here.

**Review by misaka — APPROVED** [review]

**Comment by misaka** [comment]
Looks good!

File: spec/components/task_list/view_spec.rb (hunk 10 above)
Comment (aliuk2012) [reply spec/components/task_list/view_spec.rb:4]: 100% thats is the dream, to be able to use siteprism with component tests

File: spec/features/trainees/creating_a_new_trainee_spec.rb (hunk 8 above)
Comment (aliuk2012) [reply spec/features/trainees/creating_a_new_trainee_spec.rb:41]: going to leave this in here for the moment until we get to testing this part of the journey.

**Review by edwardhorsford — APPROVED** [review]
